deep_qa/data/instances/tuple_inference_instance.py

```diff
@@ -0,0 +1,443 @@
+        """
+        Reads a TupleInstances from a line.  The format has one option (currently):
+            [question index][tab][all question+answer tuples][tab][background tuples][tab][label]
+        The question+answer tuples are formatted as:
+            [a_1-tuple_1]$$$...$$$[a_1-tuple_n]###...###[a+_m-tuple_1]$$$...$$$[a_m-tuple_p]
+        such that ``$$$`` serves as the tuple delimiter and ``###`` serves as the answer candidate
+        delimiter.
+        Each of the tuples is formatted as:
+            [s words]<>[v words]<>[o_1 words]<>...<>[o_j words]<>[context: context words]
+        such that ``<>`` serves as the tuple-slot delimiter.  Note that some slots may contain prefixes
+        indicating the type of information they contain (e.g., "context:", see below for more).
+
+        Objects are optional, and can vary in number. This makes the acceptable number of slots per
+        tuple 2 or more.
+
+        If a tuple slot is reserved for context, it contains "context:" at the start.
+
+        Some tuples contain coreference resolution info (from one of the table sources).  We do not
+        currently handle this, but plan to eventually replace the referring expression with its resolved
+        value.
+            Example: it[sky]<>is<>blue --> would become --> sky<>is<>blue
+
+        Some tuples additionally contain prefixed information about some of the "objects" such as "L:" for
+        locations and T: for time.  At present this information is simply filtered out, but can be used
+        later.
+        """
+        fields = line.split("\t")
+        if len(fields) == 4:
+            index, answers_string, background_string, label = fields
+            index = int(index)
+        elif len(fields) == 3:
+            answers_string, background_string, label = fields
+            index = None
+        else:
+            raise RuntimeError("Unrecognized line format (" + str(len(fields)) + " columns): " + line)
+
+        answers = answers_string.split("###")
+        answer_tuples = []
+        for answer in answers:
+            tuple_strings = answer.split("$$$")
+            tuples = [TextTuple.from_string(tuple_string) for tuple_string in tuple_strings if tuple_string]
+            answer_tuples.append(tuples)
+
+        background_tuple_strings = background_string.split("$$$")
+        background_tuples = [TextTuple.from_string(background_string)
+                             for background_string in background_tuple_strings if background_string]
+        label = int(label)
+        if label >= len(answer_tuples):
+            raise ConfigurationError("Invalid label, label is >= the number of answers.")
+
+        return cls(answer_tuples, background_tuples, label=label, index=index)
+
+
+class IndexedTupleInferenceInstance(IndexedInstance):
+    def __init__(self,
+                 # (num_answer_options, num_answer_tuples, num_tuple_slots, num_slot_words)
+                 answers_indexed: List,
```

Comment:
I liked having the type annotation that you had previously.  I just thought that you could add a comment _in addition_ to the type annotation.  I'd also probably put it in a docstring, or something, instead of an inline comment like this.